compiler/rustc_parse/src/parser/ty.rs

```diff
@@ -276,6 +276,34 @@ impl<'a> Parser<'a> {
     ) -> PResult<'a, Box<Ty>> {
         let allow_qpath_recovery = recover_qpath == RecoverQPath::Yes;
         maybe_recover_from_interpolated_ty_qpath!(self, allow_qpath_recovery);
+
+        if (self.token.is_keyword(kw::Const) || self.token.is_keyword(kw::Mut))
+            && self.look_ahead(1, |t| *t == token::Star)
+        {
+            let kw_span = self.token.span;
+            self.bump();
+
+            if self.eat(exp!(Star)) {
+                let star_span = self.prev_token.span;
+
+                let _ = self.parse_ty_no_question_mark_recover();
+
+                let guar = self
+                    .dcx()
+                    .struct_span_err(
+                        kw_span,
+                        "raw pointer type must have a star before the mutability keyword",
+                    )
+                    .with_multipart_suggestion(
+                        "move it to the other side",
+                        vec![(star_span, String::new()), (kw_span.shrink_to_lo(), "*".to_string())],
+                        Applicability::MachineApplicable,
+                    )
+                    .emit();
+
+                return Ok(self.mk_ty(kw_span.to(star_span), TyKind::Err(guar)));
```

Comment:
Why not recover this like a `TyKind::RawPtr`? Recovering like `TyKind::Err` seems to be strictly worse.